File: backend/src/pomelo_orbit/interfaces/api/cd/route.py
"""路由管理 API"""

import logging
import math
from typing import Annotated

from fastapi import APIRouter, Depends, Query, UploadFile

from pomelo_orbit.application.cd.di import get_route_service
from pomelo_orbit.application.cd.route_service import RouteService
from pomelo_orbit.application.project.di import get_project_service
from pomelo_orbit.application.project.project_service import ProjectService
from pomelo_orbit.domain.auth.entities import User
from pomelo_orbit.domain.cd.entities import Route
from pomelo_orbit.domain.exceptions import BusinessError
from pomelo_orbit.interfaces.api.auth.dependencies import get_current_user
from pomelo_orbit.interfaces.api.cd.dto.route import RouteCreateReq, RouteResp, RouteUpdateReq
from pomelo_orbit.interfaces.api.common import PaginatedResp

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/route", tags=["route"])


def _authorize_route_project(
    route_service: RouteService,
    project_service: ProjectService,
    current_user: User,
    route_id: str,
) -> Route:
    route = route_service.get_route(route_id)
    if route.project_id is None:
        raise BusinessError("Permission denied", status_code=403)
    project_service.get_project(current_user.id, route.project_id)
    return route


@router.get("", response_model=PaginatedResp[RouteResp])
def list_routes(
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
    project_id: Annotated[str, Query()],
    page: Annotated[int, Query(ge=1)] = 1,
    per_page: Annotated[int, Query(ge=1, le=100)] = 10,
    search: Annotated[str | None, Query()] = None,
) -> PaginatedResp[RouteResp]:
    """列出所有路由"""
    project_service.get_project(current_user.id, project_id)
    routes, total = route_service.list_routes(project_id, page, per_page, search)
    return PaginatedResp(
        items=[RouteResp.model_validate(r) for r in routes],
        total=total,
        page=page,
        per_page=per_page,
        pages=math.ceil(total / per_page) if total > 0 else 1,
    )


@router.post("", response_model=RouteResp, status_code=201)
def create_route(
    data: RouteCreateReq,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
    project_id: Annotated[str, Query()],
) -> RouteResp:
    """创建路由"""
    project_service.get_project(current_user.id, project_id)
    route = route_service.create_route(
        project_id=project_id,
        name=data.name,
        domain=data.domain,
        path_prefix=data.path_prefix,
        target_url=data.target_url,
        enabled=data.enabled,
    )
    return RouteResp.model_validate(route)


@router.get("/{route_id}", response_model=RouteResp)
def get_route(
    route_id: str,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> RouteResp:
    """获取路由详情"""
    route = _authorize_route_project(route_service, project_service, current_user, route_id)
    return RouteResp.model_validate(route)


@router.put("/{route_id}", response_model=RouteResp)
def update_route(
    route_id: str,
    data: RouteUpdateReq,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> RouteResp:
    """更新路由"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    update_data = data.model_dump(exclude_unset=True)
    updated_route = route_service.update_route(route_id, **update_data)
    return RouteResp.model_validate(updated_route)


@router.delete("/{route_id}", status_code=204)
def delete_route(
    route_id: str,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """删除路由"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    route_service.delete_route(route_id)


@router.post("/{route_id}/enable")
def enable_route(
    route_id: str,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """启用路由"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    route_service.enable_route(route_id)
    return {"message": "Route enabled successfully"}


@router.post("/{route_id}/disable")
def disable_route(
    route_id: str,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """停用路由"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    route_service.disable_route(route_id)
    return {"message": "Route disabled successfully"}


@router.post("/sync")
def sync_routes(
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
    project_id: Annotated[str, Query()],
):
    """同步路由配置"""
    project_service.get_project(current_user.id, project_id)
    route_service.sync_routes(project_id)
    return {"message": "Routes synced successfully"}


@router.post("/{route_id}/cert")
async def upload_cert(
    route_id: str,
    pem: UploadFile,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> RouteResp:
    """上传 SSL 证书"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    cert_content = await pem.read()
    route = route_service.upload_cert(route_id, cert_content)
    return RouteResp.model_validate(route)


@router.delete("/{route_id}/https")
def disable_https(
    route_id: str,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> RouteResp:
    """禁用 HTTPS"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    route = route_service.disable_https(route_id)
    return RouteResp.model_validate(route)


@router.post("/{route_id}/letsencrypt")
def enable_letsencrypt(
    route_id: str,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> RouteResp:
    """启用 Let's Encrypt 自动证书"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    route = route_service.enable_letsencrypt(route_id)
    return RouteResp.model_validate(route)


@router.post("/{route_id}/mkcert")
def enable_mkcert(
    route_id: str,
    route_service: Annotated[RouteService, Depends(get_route_service)],
    project_service: Annotated[ProjectService, Depends(get_project_service)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> RouteResp:
    """启用 mkcert 本地证书"""
    _authorize_route_project(route_service, project_service, current_user, route_id)
    route = route_service.enable_mkcert(route_id)
    return RouteResp.model_validate(route)

File: internal/api/http/handler/cd/route.go
package cdhandler

import (
	"bytes"
	"encoding/pem"
	"net/http"
	"strings"

	transportcodec "gitee.com/leoninew/PomeloOrbit-go/internal/api/http/codec"
	pomeloorbit "gitee.com/leoninew/PomeloOrbit-go/internal/gen/proto/orbit/v1"
	"github.com/gin-gonic/gin"

	transportresponse "gitee.com/leoninew/PomeloOrbit-go/internal/api/http/response"
	cddto "gitee.com/leoninew/PomeloOrbit-go/internal/application/cd/dto"
	"gitee.com/leoninew/PomeloOrbit-go/internal/model"
)

func (h Handler) RegisterRouteRoutes(r router) {
	r.GET("/api/cd/route", h.listRoutes)
	r.POST("/api/cd/route", h.createRoute)
	r.POST("/api/cd/route/sync", h.syncRoutes)
	r.GET("/api/cd/route/:route_id", h.getRoute)
	r.PUT("/api/cd/route/:route_id", h.updateRoute)
	r.DELETE("/api/cd/route/:route_id", h.deleteRoute)
	r.POST("/api/cd/route/:route_id/enable", h.enableRoute)
	r.POST("/api/cd/route/:route_id/disable", h.disableRoute)
	r.POST("/api/cd/route/:route_id/cert", h.uploadRouteCert)
	r.DELETE("/api/cd/route/:route_id/https", h.disableRouteHTTPS)
	r.POST("/api/cd/route/:route_id/letsencrypt", h.enableRouteLetsEncrypt)
	r.POST("/api/cd/route/:route_id/mkcert", h.enableRouteMkcert)
}

func (h Handler) RegisterTraefikRouteRoutes(r router) {
	r.GET("/api/cd/traefik-route/config", h.getTraefikRouteConfig)
	r.GET("/api/cd/traefik-route", h.listTraefikRoutes)
}

func (h Handler) listRoutes(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	page := transportresponse.QueryInt(c.Request.URL.Query().Get("page"), 1)
	perPage := transportresponse.QueryInt(c.Request.URL.Query().Get("per_page"), 10)
	items, err := h.service.ListRoutes(c.Request.Context(), current.Id, c.Request.URL.Query().Get("project_id"), page, perPage, c.Request.URL.Query().Get("search"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := mapPage(items, routeResponse)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &pomeloorbit.RoutePaginatedResp{Items: transportresponse.Ptrs(resp.Items), Total: int32(resp.Total), Page: int32(resp.Page), PerPage: int32(resp.PerPage), Pages: int32(transportresponse.PageCount(resp.Total, resp.PerPage))}})
}

func (h Handler) createRoute(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.RouteCreateReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	route, err := h.service.CreateRoute(c.Request.Context(), current.Id, c.Request.URL.Query().Get("project_id"), cddto.RouteCreateInput{Name: req.Name, Domain: req.Domain, PathPrefix: req.PathPrefix, TargetURL: req.TargetUrl, Enabled: req.Enabled})
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := routeResponse(route)
	c.Render(http.StatusCreated, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) getRoute(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	route, err := h.service.RouteForUser(c.Request.Context(), current.Id, c.Param("route_id"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := routeResponse(route)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) updateRoute(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.RouteUpdateReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	route, err := h.service.UpdateRoute(c.Request.Context(), current.Id, c.Param("route_id"), cddto.RouteUpdateInput{Name: req.Name, Domain: req.Domain, PathPrefix: req.PathPrefix, TargetURL: req.TargetUrl, Enabled: req.Enabled})
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := routeResponse(route)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) deleteRoute(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	if err := h.service.DeleteRoute(c.Request.Context(), current.Id, c.Param("route_id")); err != nil {
		h.writeError(c, err)
		return
	}
	c.Status(http.StatusNoContent)
}

func (h Handler) enableRoute(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.RouteEnableReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	if _, err := h.service.EnableRoute(c.Request.Context(), current.Id, c.Param("route_id")); err != nil {
		h.writeError(c, err)
		return
	}
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &pomeloorbit.RouteEnableResp{Message: "Route enabled successfully"}})
}

func (h Handler) disableRoute(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.RouteDisableReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	if _, err := h.service.DisableRoute(c.Request.Context(), current.Id, c.Param("route_id")); err != nil {
		h.writeError(c, err)
		return
	}
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &pomeloorbit.RouteDisableResp{Message: "Route disabled successfully"}})
}

func (h Handler) syncRoutes(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.RouteSyncReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	if err := h.service.SyncRoutes(c.Request.Context(), current.Id, c.Request.URL.Query().Get("project_id")); err != nil {
		h.writeError(c, err)
		return
	}
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &pomeloorbit.RouteSyncResp{Message: "Routes synced successfully"}})
}

func (h Handler) uploadRouteCert(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	file, _, err := c.Request.FormFile("pem")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "pem is required"})
		return
	}
	defer func() { _ = file.Close() }()
	var content bytes.Buffer
	if _, err := content.ReadFrom(file); err != nil {
		h.logger.Error("read certificate upload failed", "route_id", c.Param("route_id"), "error", err)
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "Failed to read certificate"})
		return
	}
	certPEM, certKey, ok := splitPEM(content.Bytes())
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid PEM certificate"})
		return
	}
	route, err := h.service.UploadRouteCert(c.Request.Context(), current.Id, c.Param("route_id"), certPEM, certKey)
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := routeResponse(route)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) disableRouteHTTPS(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	route, err := h.service.DisableRouteHTTPS(c.Request.Context(), current.Id, c.Param("route_id"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := routeResponse(route)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) enableRouteLetsEncrypt(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.RouteLetsEncryptEnableReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	route, err := h.service.EnableRouteLetsEncrypt(c.Request.Context(), current.Id, c.Param("route_id"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := routeResponse(route)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) enableRouteMkcert(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.RouteMkcertEnableReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	route, err := h.service.EnableRouteMkcert(c.Request.Context(), current.Id, c.Param("route_id"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := routeResponse(route)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) getTraefikRouteConfig(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	config, err := h.service.TraefikRouteConfig(c.Request.Context(), current.Id, c.Request.URL.Query().Get("project_id"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := traefikConfigResponse(config)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) listTraefikRoutes(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	items, err := h.service.ListTraefikRoutes(c.Request.Context(), current.Id, c.Request.URL.Query().Get("project_id"))
	if err != nil {
		if strings.HasPrefix(err.Error(), "无法连接到 Traefik:") {
			c.JSON(http.StatusServiceUnavailable, gin.H{"detail": err.Error()})
			return
		}
		h.writeError(c, err)
		return
	}
	resp := traefikRouteListResponse(items)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func routeResponse(route model.Route) pomeloorbit.RouteResp {
	return pomeloorbit.RouteResp{Id: route.Id, Name: route.Name, Domain: route.Domain, PathPrefix: route.PathPrefix, TargetUrl: route.TargetURL, Enabled: route.Enabled, HttpsEnabled: route.HTTPSEnabled, CertType: route.CertType, CreatedAt: transportresponse.FormatTime(route.CreatedAt), UpdatedAt: transportresponse.FormatTime(route.UpdatedAt)}
}

func traefikConfigResponse(config cddto.TraefikConfigResp) pomeloorbit.TraefikConfigResp {
	return pomeloorbit.TraefikConfigResp{DashboardDomain: config.DashboardDomain, HttpsEnabled: config.HTTPSEnabled}
}

func traefikRouteListResponse(resp cddto.TraefikRouteListResp) pomeloorbit.TraefikRouteListResp {
	items := make([]pomeloorbit.TraefikRouterResp, 0, len(resp.Items))
	for _, item := range resp.Items {
		items = append(items, traefikRouterResponse(item))
	}
	return pomeloorbit.TraefikRouteListResp{Items: transportresponse.Ptrs(items), Total: int32(resp.Total)}
}

func traefikRouterResponse(router cddto.TraefikRouterResp) pomeloorbit.TraefikRouterResp {
	return pomeloorbit.TraefikRouterResp{Name: router.Name, Provider: router.Provider, Status: router.Status, Rule: router.Rule, Service: router.Service, Entrypoints: append([]string(nil), router.Entrypoints...), Tls: router.TLS}
}

func splitPEM(content []byte) (string, string, bool) {
	var cert bytes.Buffer
	var key bytes.Buffer
	remaining := content
	for {
		block, rest := pem.Decode(remaining)
		if block == nil {
			break
		}
		encoded := pem.EncodeToMemory(block)
		if block.Type == "CERTIFICATE" {
			cert.Write(encoded)
		} else if strings.Contains(block.Type, "PRIVATE KEY") {
			key.Write(encoded)
		}
		if len(rest) == len(remaining) {
			break
		}
		remaining = rest
	}
	if cert.Len() == 0 || key.Len() == 0 {
		return "", "", false
	}
	return cert.String(), key.String(), true
}
